### PR TITLE
HHH-16515 - Add @Nullable annotations to org.hibernate.spi

### PR DESCRIPTION
**Pull request by jrenaat**

### CODE REVIEW

File: gradle/java-module.gradle [hunk 1]
```diff
@@ -523,7 +523,7 @@ checkerFramework {
 	extraJavacArgs = [
 	        '-AsuppressWarnings=initialization',
 			"-Astubs=${project.rootDir}/checkerstubs",
-			'-AonlyDefs=^org\\.hibernate\\.jpamodelgen\\.'
+			'-AonlyDefs=^org\\.hibernate\\.jpamodelgen\\.,^org\\.hibernate\\.spi\\.'
```
Comment (beikov) [inline gradle/java-module.gradle:526]: AFAICS from the documentation, the whole argument is a single regex, so you'd need this instead:

```suggestion
			'-AonlyDefs=^org\\.hibernate\\.(jpamodelgen|spi)\\.'
```

**Review by beikov — COMMENTED** [review]
Other than what I commented on, this looks good to me. My suggested change will probably also fix the problems you have been running into.

File: hibernate-core/src/main/java/org/hibernate/spi/DotIdentifierSequence.java [hunk 2]
```diff
@@ -26,6 +28,7 @@ public interface DotIdentifierSequence {
 	 * Given the sequence {@code a.b.c}, returns the sequence
 	 * {@code a.b}.
 	 */
+	@Nullable
 	DotIdentifierSequence getParent();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/DotIdentifierSequence.java:32]: The checker framework guideline is to add the annotation to the type. Please use this:

```suggestion
	@Nullable DotIdentifierSequence getParent();
```

File: hibernate-core/src/main/java/org/hibernate/spi/DotIdentifierSequence.java [hunk 3]
```diff
@@ -65,7 +68,7 @@ default DotIdentifierSequence[] getParts() {
 
 	private void parts(List<DotIdentifierSequence> list) {
 		if ( getParent() != null ) {
-			getParent().parts( list );
+			NullnessUtil.castNonNull( getParent() ).parts( list );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/DotIdentifierSequence.java:71]:
```suggestion
		DotIdentifierSequence parent = getParent();
		if ( parent != null ) {
			parent .parts( list );
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 4]
```diff
@@ -49,10 +52,10 @@ public NavigablePath(NavigablePath parent, String navigableName) {
 		this( parent, navigableName, null );
 	}
 
-	public NavigablePath(NavigablePath parent, String localName, String alias) {
+	public NavigablePath(NavigablePath parent, String localName, @Nullable String alias) {
 		assert parent != null;
 
-		this.parent = parent;
+		this.parent = NullnessUtil.castNonNull( parent );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:58]: This change shouldn't be necessary.

```suggestion
		this.parent = parent;
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 5]
```diff
@@ -323,24 +326,24 @@ public String toString() {
...
+	protected static String calculateRootFullPath(@Nullable NavigablePath parent, String rootName, @Nullable String alias) {
 		assert parent == null;
 		return alias == null ? rootName : rootName + "(" + alias + ")";
 	}
 
 	/**
 	 * The normal pattern used for the "full path"
 	 */
-	private static String calculateNormalFullPath(NavigablePath parent, String localName, String alias) {
+	private static String calculateNormalFullPath(@Nullable NavigablePath parent, String localName, @Nullable String alias) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:343]:
```suggestion
	private static String calculateNormalFullPath(NavigablePath parent, String localName, @Nullable String alias) {
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 6]
```diff
@@ -323,24 +326,24 @@ public String toString() {
...
 		assert parent != null;
 
-		final String parentFullPath = parent.getFullPath();
+		final String parentFullPath = NullnessUtil.castNonNull( parent ).getFullPath();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:346]:
```suggestion
		final String parentFullPath = parent.getFullPath();
```

File: hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java [hunk 7]
```diff
@@ -32,7 +34,7 @@ public TreatedNavigablePath(NavigablePath parent, String entityTypeName, String
 		assert !( parent instanceof TreatedNavigablePath );
 	}
 
-	protected static String calculateTreatedFullPath(NavigablePath parent, String localName, String alias) {
+	protected static String calculateTreatedFullPath(@Nullable NavigablePath parent, String localName, @Nullable String alias) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java:37]:
```suggestion
	protected static String calculateTreatedFullPath(NavigablePath parent, String localName, @Nullable String alias) {
```

File: hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java [hunk 8]
```diff
@@ -16,11 +18,11 @@
 @Incubating
 public class TreatedNavigablePath extends NavigablePath {
 
-	public TreatedNavigablePath(NavigablePath parent, String entityTypeName) {
+	public TreatedNavigablePath(@Nullable NavigablePath parent, String entityTypeName) {
 		this( parent, entityTypeName, null );
 	}
 
-	public TreatedNavigablePath(NavigablePath parent, String entityTypeName, String alias) {
+	public TreatedNavigablePath(@Nullable NavigablePath parent, String entityTypeName, @Nullable String alias) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java:25]:
```suggestion
	public TreatedNavigablePath(NavigablePath parent, String entityTypeName, @Nullable String alias) {
```

File: hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java [hunk 9]
```diff
@@ -16,11 +18,11 @@
 @Incubating
 public class TreatedNavigablePath extends NavigablePath {
 
-	public TreatedNavigablePath(NavigablePath parent, String entityTypeName) {
+	public TreatedNavigablePath(@Nullable NavigablePath parent, String entityTypeName) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/TreatedNavigablePath.java:21]:
```suggestion
	public TreatedNavigablePath(NavigablePath parent, String entityTypeName) {
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 10]
```diff
@@ -216,23 +220,23 @@ public boolean isSuffix(DotIdentifierSequence dotIdentifierSequence) {
...
+	public @Nullable NavigablePath trimSuffix(@Nullable DotIdentifierSequence suffix) {
 		if ( suffix == null ) {
 			return this;
 		}
 		if ( !getLocalName().equals( suffix.getLocalName() ) ) {
 			return null;
 		}
 		if ( getParent() != null ) {
-			return getParent().trimSuffix( suffix.getParent() );
+			return castNonNull( getParent() ).trimSuffix( suffix.getParent() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:231]:
```suggestion
		DotIdentifierSequence parent = getParent();
		if ( parent  != null ) {
			return parent.trimSuffix( suffix.getParent() );
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 11]
```diff
@@ -195,14 +199,14 @@ public boolean isParent(NavigablePath navigablePath) {
...
-	public boolean isSuffix(DotIdentifierSequence dotIdentifierSequence) {
+	public boolean isSuffix(@Nullable DotIdentifierSequence dotIdentifierSequence) {
 		if ( dotIdentifierSequence == null ) {
 			return true;
 		}
 		if ( !localNamesMatch( dotIdentifierSequence ) ) {
 			return false;
 		}
-		return getParent() != null && getParent().isSuffix( dotIdentifierSequence.getParent() );
+		return getParent() != null && castNonNull( getParent() ).isSuffix( dotIdentifierSequence.getParent() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:209]:
```suggestion
		DotIdentifierSequence parent = getParent();
		return parent  != null && parent.isSuffix( dotIdentifierSequence.getParent() );
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java [hunk 12]
```diff
@@ -323,21 +327,21 @@ public String toString() {
 	 */
 	@FunctionalInterface
 	protected interface FullPathCalculator extends Serializable {
-		String calculateFullPath(NavigablePath parent, String localName, String alias);
+		String calculateFullPath(NavigablePath parent, String localName, @Nullable String alias);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:330]: You were right about this.

```suggestion
		String calculateFullPath(@Nullable NavigablePath parent, String localName, @Nullable String alias);
```

File: hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java (hunk 11 above)
Comment (jrenaat) [reply hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:209]: The reference needs to be a NavigablePath I think?

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/spi/NavigablePath.java:209]: yeah, you're right

**Review by beikov — APPROVED** [review]
Please also backport to 6.2

**Pull request merged**